src/accountsdb/db.zig

```diff
@@ -808,263 +986,1058 @@ pub const AccountsDB = struct {
+        switch (account_ref.location) {
+            .File => |ref_info| {
+                const account = try self.getAccountInFile(
+                    ref_info.file_id,
+                    ref_info.offset,
+                );
+
+                return account;
+            },
+            .Cache => |ref_info| {
+                const account_cache, var account_cache_lg = self.account_cache.readWithLock();
+                defer account_cache_lg.unlock();
+
+                _, const accounts = account_cache.get(account_ref.slot) orelse return error.SlotNotFound;
+                const account = accounts[ref_info.index];
+
+                return account.clone(self.allocator);
+            },
+        }
+    }
+
+    /// gets an account given an file_id and offset value
+    pub fn getAccountInFile(
+        self: *Self,
+        file_id: FileId,
         offset: usize,
-    ) !AccountInFile {
-        const accounts_file: AccountFile = self.getAccountFile(file_id) orelse {
-            return error.FileIdNotFound;
+    ) !Account {
+        var account_file_rw: RwMux(AccountFile) = blk: {
+            const file_map, var file_map_lg = self.file_map.readWithLock();
+            defer file_map_lg.unlock();
+
+            break :blk file_map.get(file_id) orelse {
+                return error.FileIdNotFound;
+            };
         };
-        const account = accounts_file.readAccount(offset) catch {
+
+        const account_file, var account_file_lg = account_file_rw.readWithLock();
+        defer account_file_lg.unlock();
+
+        const account_in_file = account_file.readAccount(offset) catch {
             return error.InvalidOffset;
         };
+        const account = account_in_file.toAccount().clone(self.allocator);
         return account;
     }
+
+    pub fn getAccountHashAndLamportsFromRef(
+        self: *Self,
+        account_ref: *const AccountRef,
+    ) !struct { hash: Hash, lamports: u64 } {
+        switch (account_ref.location) {
+            .File => |ref_info| {
+                var account_file_rw = blk: {
+                    const file_map, var file_map_lg = self.file_map.readWithLock();
+                    defer file_map_lg.unlock();
+
+                    break :blk file_map.get(ref_info.file_id) orelse {
+                        return error.FileIdNotFound;
+                    };
+                };
+
+                const account_file, var account_file_lg = account_file_rw.readWithLock();
+                defer account_file_lg.unlock();
+
+                const result = account_file.getAccountHashAndLamports(
+                    ref_info.offset,
+                ) catch return error.InvalidOffset;
+
+                return .{
+                    .hash = result.hash.*,
+                    .lamports = result.lamports.*,
+                };
+            },
+            .Cache => |_| {
+                // we dont use this method for cache
+                return error.NotImplemented;
```

Comment:
Why have this extra branch hidden inside the function? This method doesn't interact with any of the other fields in the account_ref, it'd make more sense to me to just accept the payload type of `AccountLocation.File` as a parameter (it's currently an anonymous declaration, would need to give it a name).

Reply:
changed to take location as param - 94aabe7

rn we dont use the method for cache but its _very likely_ we will use it in the future so i think its ok to leave it like this for now

Reply:
I think that's fine then if it is eventually intended, but in that case I think it would be better to mark this as a panic, to indicate that it is unimplemented, and not just a normal failure state that is reachable assuming normal and correct use of the API (the error set is part of the API after all).

Reply:
thats a good point - 87d2d9b